pkg/streamingpromql/operators/functions/rate_increase.go

```diff
@@ -135,18 +145,24 @@ func histogramRate(isRate bool, step types.RangeVectorStepData, hHead []promql.H
 	err = accumulate(hTail)
 	if err != nil {
 		return nil, err
+	}
 
+	if !isCounter && (firstPoint.H.CounterResetHint != histogram.GaugeType || lastPoint.H.CounterResetHint != histogram.GaugeType) {
+		emitAnnotation(annotations.NewNativeHistogramNotGaugeWarning)
 	}
 
 	if delta.Schema != desiredSchema {
 		delta = delta.CopyToSchema(desiredSchema)
 	}
 
-	val := calculateHistogramRate(isRate, step.RangeStart, step.RangeEnd, rangeSeconds, firstPoint, lastPoint, delta, hCount)
-	return val, err
+	if isCounter {
+		val := calculateHistogramRate(isRate, step.RangeStart, step.RangeEnd, rangeSeconds, firstPoint, lastPoint, delta, hCount)
+		return val, err
+	}
+	return delta, err
```

Comment:
I'm not sure this is quite right - [the docs for `delta`](https://prometheus.io/docs/prometheus/latest/querying/functions/#delta) say:

> `delta(v range-vector)` calculates the difference between the first and last value of each time series element in a range vector `v`, returning an instant vector with the given deltas and equivalent labels. The delta is extrapolated to cover the full time range as specified in the range vector selector...

I see two issues with the implementation here:

* we're not doing extrapolation
* we're considering resets in `delta`, but we want to consider just the first and last value

I wonder if it would be clearer and simpler for `delta` to not use `rate` above, and instead be implemented separately and call `calculateFloatRate` and `calculateHistogramRate` to perform extrapolation.

Reply:
Actually I was looking at how prometheus engine was implementing delta. But let me check at this again and try to apply your suggestion.